File: src/api/routes/v1/dto/user_preferences.rs
//! User Preferences DTOs

use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};
use std::collections::HashMap;
use utoipa::ToSchema;

use crate::db::entities::user_preferences;
use crate::db::repositories::UserPreferencesRepository;

/// A single user preference
#[derive(Debug, Clone, Serialize, Deserialize, ToSchema)]
#[serde(rename_all = "camelCase")]
pub struct UserPreferenceDto {
    /// The preference key (e.g., "ui.theme", "reader.zoom")
    #[schema(example = "ui.theme")]
    pub key: String,

    /// The preference value
    #[schema(example = json!("dark"))]
    pub value: serde_json::Value,

    /// The value type: string, integer, float, boolean, or json
    #[schema(example = "string")]
    pub value_type: String,

    /// When the preference was last updated
    #[schema(example = "2024-01-15T18:45:00Z")]
    pub updated_at: DateTime<Utc>,
}

impl TryFrom<user_preferences::Model> for UserPreferenceDto {
    type Error = anyhow::Error;

    fn try_from(model: user_preferences::Model) -> Result<Self, Self::Error> {
        let value = UserPreferencesRepository::to_json_value(&model)?;
        Ok(Self {
            key: model.key,
            value,
            value_type: model.value_type,
            updated_at: model.updated_at,
        })
    }
}

/// Response containing all user preferences
#[derive(Debug, Clone, Serialize, Deserialize, ToSchema)]
#[serde(rename_all = "camelCase")]
pub struct UserPreferencesResponse {
    /// List of all user preferences
    pub preferences: Vec<UserPreferenceDto>,
}

/// Request to set a single preference value
#[derive(Debug, Clone, Serialize, Deserialize, ToSchema)]
#[serde(rename_all = "camelCase")]
pub struct SetPreferenceRequest {
    /// The value to set
    #[schema(example = json!("dark"))]
    pub value: serde_json::Value,
}

/// Request to set multiple preferences at once
#[derive(Debug, Clone, Serialize, Deserialize, ToSchema)]
#[serde(rename_all = "camelCase")]
pub struct BulkSetPreferencesRequest {
    /// Map of preference keys to values
    #[schema(example = json!({"ui.theme": "dark", "reader.zoom": 150}))]
    pub preferences: HashMap<String, serde_json::Value>,
}

/// Response after setting preferences
#[derive(Debug, Clone, Serialize, Deserialize, ToSchema)]
#[serde(rename_all = "camelCase")]
pub struct SetPreferencesResponse {
    /// Number of preferences that were updated
    #[schema(example = 3)]
    pub updated: usize,

    /// The updated preferences
    pub preferences: Vec<UserPreferenceDto>,
}

/// Response after deleting a preference
#[derive(Debug, Clone, Serialize, Deserialize, ToSchema)]
#[serde(rename_all = "camelCase")]
pub struct DeletePreferenceResponse {
    /// Whether a preference was deleted
    #[schema(example = true)]
    pub deleted: bool,

    /// Message describing the result
    #[schema(example = "Preference 'ui.theme' was reset to default")]
    pub message: String,
}

File: web/src/api/userPreferences.ts
import type { components } from "@/types/api.generated";
import type { PreferenceKey, TypedPreferences } from "@/types/preferences";
import { api } from "./client";

// Re-export generated types for convenience
export type UserPreferenceDto = components["schemas"]["UserPreferenceDto"];
export type UserPreferencesResponse =
  components["schemas"]["UserPreferencesResponse"];
export type SetPreferenceRequest =
  components["schemas"]["SetPreferenceRequest"];
export type BulkSetPreferencesRequest =
  components["schemas"]["BulkSetPreferencesRequest"];
export type SetPreferencesResponse =
  components["schemas"]["SetPreferencesResponse"];

export const userPreferencesApi = {
  /**
   * Get all user preferences
   */
  getAll: async (): Promise<UserPreferenceDto[]> => {
    const response =
      await api.get<UserPreferencesResponse>("/user/preferences");
    return response.data.preferences;
  },

  /**
   * Get a single preference by key
   */
  get: async <K extends PreferenceKey>(
    key: K,
  ): Promise<UserPreferenceDto | null> => {
    try {
      const response = await api.get<UserPreferenceDto>(
        `/user/preferences/${encodeURIComponent(key)}`,
      );
      return response.data;
    } catch (error) {
      // Return null if preference not found (404)
      if (
        error &&
        typeof error === "object" &&
        "response" in error &&
        (error as { response?: { status?: number } }).response?.status === 404
      ) {
        return null;
      }
      throw error;
    }
  },

  /**
   * Set a single preference
   */
  set: async <K extends PreferenceKey>(
    key: K,
    value: TypedPreferences[K],
  ): Promise<UserPreferenceDto> => {
    const response = await api.put<UserPreferenceDto>(
      `/user/preferences/${encodeURIComponent(key)}`,
      { value } as SetPreferenceRequest,
    );
    return response.data;
  },

  /**
   * Bulk set multiple preferences at once
   */
  bulkSet: async (
    preferences: Partial<TypedPreferences>,
  ): Promise<SetPreferencesResponse> => {
    const response = await api.put<SetPreferencesResponse>(
      "/user/preferences",
      { preferences } as BulkSetPreferencesRequest,
    );
    return response.data;
  },

  /**
   * Delete (reset) a preference to its default value
   */
  delete: async (key: PreferenceKey): Promise<void> => {
    await api.delete(`/user/preferences/${encodeURIComponent(key)}`);
  },
};
